tower-http/src/catch_panic.rs

```diff
@@ -0,0 +1,202 @@
+}
+
+impl<S> CatchPanic<S> {
+    /// Create a new `CatchPanic`.
+    pub fn new(inner: S) -> Self {
+        Self { inner }
+    }
+
+    define_inner_service_accessors!();
+
+    /// Returns a new [`Layer`] that wraps services with a `CatchPanic` middleware.
+    ///
+    /// [`Layer`]: tower_layer::Layer
+    pub fn layer() -> CatchPanicLayer {
+        CatchPanicLayer::new()
+    }
+}
+
+impl<S, ReqBody, ResBody> Service<Request<ReqBody>> for CatchPanic<S>
+where
+    S: Service<Request<ReqBody>, Response = Response<ResBody>>,
+    ResBody: Body<Data = Bytes> + Send + 'static,
+{
+    type Response = Response<UnsyncBoxBody<Bytes, ResBody::Error>>;
+    type Error = S::Error;
+    type Future = ResponseFuture<S::Future>;
+
+    #[inline]
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request<ReqBody>) -> Self::Future {
+        match std::panic::catch_unwind(AssertUnwindSafe(|| self.inner.call(req))) {
+            Ok(future) => ResponseFuture {
+                kind: Kind::Future {
+                    future: AssertUnwindSafe(future).catch_unwind(),
+                },
+            },
+            Err(panic_err) => ResponseFuture {
+                kind: Kind::Panicked {
+                    panic_err: Some(panic_err),
+                },
+            },
+        }
+    }
+}
+
+pin_project! {
+    /// Response future for [`CatchPanic`].
+    pub struct ResponseFuture<F> {
+        #[pin]
+        kind: Kind<F>,
+    }
+}
+
+pin_project! {
+    #[project = KindProj]
+    enum Kind<F> {
+        Panicked {
+            panic_err: Option<Box<dyn Any + Send + 'static>>,
+        },
+        Future {
+            #[pin]
+            future: CatchUnwind<AssertUnwindSafe<F>>,
+        }
+    }
+}
+
+impl<F, ResBody, E> Future for ResponseFuture<F>
+where
+    F: Future<Output = Result<Response<ResBody>, E>>,
+    ResBody: Body<Data = Bytes> + Send + 'static,
+{
+    type Output = Result<Response<UnsyncBoxBody<Bytes, ResBody::Error>>, E>;
+
+    #[allow(warnings)]
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        match self.project().kind.project() {
+            KindProj::Panicked { panic_err } => Poll::Ready(Ok(panic_err_to_response(
+                panic_err.take().expect("future polled after completion"),
+            ))),
+            KindProj::Future { future } => match ready!(future.poll(cx)) {
+                Ok(Ok(res)) => Poll::Ready(Ok(res.map(|body| body.boxed_unsync()))),
+                Ok(Err(svc_err)) => Poll::Ready(Err(svc_err)),
+                Err(panic_err) => Poll::Ready(Ok(panic_err_to_response(panic_err))),
+            },
+        }
+    }
+}
+
+fn panic_err_to_response<E>(
+    err: Box<dyn Any + Send + 'static>,
+) -> Response<UnsyncBoxBody<Bytes, E>> {
+    let message: Cow<_> = if let Some(s) = err.downcast_ref::<String>() {
+        s.to_owned().into()
```

Comment:
Microoptimization: Since we're owning and thowing away the `Box` anyway, we might do a `downcast_mut()` and `swap()`ing with an empty `String`? (saves the `to_owned()`)